contentprovider.go

```diff
@@ -392,6 +393,41 @@ func chunkCandidates(ms []*candidateMatch, newlines newlines, numContextLines in
 	return chunks
 }
 
+// columnHelper is a helper struct which caches the number of runes last
+// counted. If we naively use utf8.RuneCount for each match on a line, this
+// leads to an O(nm) algorithm where m is the number of matches and n is the
+// length of the line. Aassuming we our candidates are increasing in offset
```

Comment:
```suggestion
// length of the line. Assuming we our candidates are increasing in offset
```

Reply:
Just to check: are we always sure we can assume our candidates are increasing in offset? I can't remember if this is always true.

Reply:
Oh, reading the implementation, I guess we just fall back to the less performant version.

Reply:
I believe this invariant is true, because in `gatherMatches` we always make sure to sort by `byteOffset`.

Maybe we could update the comments to make it clear this invariant is assumed, and treat the unsorted case as an error rather than being expected? That way if we ever introduce a bug here, we don't silently fall back to an O(n^2) algorithm... much harder to track down then a clear error in testing.

General thought: if invariants are too tricky to reason about, sometimes I just explicitly add a (re)sort! I believe Go's default sort is very fast when the input is already sorted. This bounds the worst case nicely.

Reply:
I also checked the invariants. We also have the invariant that things don't overlap, which is also important since we lookup the end column.

The sorted invariant is actually quite important for other bits of code like chunkCandidates. So what I did was add a sorted check which loudly complains and then sorts if the invariant is broken.

Initially I pretended I was a haskell programmer and added a special type which guaranteed this, but TBH it felt quite overengineered. Happy to try it out if there is interest, but for now gonna merge with extra perf invariant documentation and sort check.